src/Clash/Cores/Etherbone/Base.hs

```diff
@@ -0,0 +1,291 @@
+etherbonePacketizerC = packetizerC (const ()) metaMap
+  where
+    metaMap :: EBHeader -> EBHeader
+    metaMap hdr = hdr
+
+-- | Strips 4 bytes from the start of a packet in the case of a 64-bit wishbone
+-- bus and passes the input for a 32-bit bus.
+-- This is needed to get rid of the 'potential padding' in the case of a 64-bit
+-- bus.
+etherbonePaddingStripperC :: forall dom dataWidth .
+  ( HiddenClockResetEnable dom
+  , KnownNat dataWidth
+  , 4 <= dataWidth
+  )
+  => Circuit (PacketStream dom dataWidth EBHeader) (PacketStream dom dataWidth EBHeader)
+etherbonePaddingStripperC = case compareSNat (SNat @dataWidth) d4 of
+  SNatLE -> idC
+  SNatGT -> depacketizerC metaMap
+  where
+    metaMap :: Vec (dataWidth-4) (BitVector 8) -> EBHeader -> EBHeader
+    metaMap _ hdr = hdr
+
+-- | Adds 4 bytes of padding to the start of a packet in the case of a 64-bit
+-- wishbone bus. It passes the input for a 32-bit bus.
+-- Needed to apply the 'potential padding' in the case of a 64-bit wishbone bus.
+etherbonePaddingAdderC :: forall dom dataWidth .
+  ( HiddenClockResetEnable dom
+  , KnownNat dataWidth
+  , 4 <= dataWidth
+  )
+  => Circuit (PacketStream dom dataWidth EBHeader) (PacketStream dom dataWidth EBHeader)
+etherbonePaddingAdderC = case compareSNat (SNat @dataWidth) d4 of
+  SNatLE -> idC
+  SNatGT -> packetizerC id padMap
+  where
+    padMap :: EBHeader -> Vec (dataWidth-4) (BitVector 8)
+    padMap _ = repeat 0
+
+-- | Extract 'RecordHeader' data from a @PacketStream@ into the metadata. Strips
+-- potential padding from the header. Splits a Etherbone packet @PacketStream@
+-- into separate Record @PacketStream@s. The added bool in the metadata
+-- indicates end-of-packet.
+--
+-- 'EBHeader' can be dropped. All the info in it is static and known. 
+recordDepacketizerC :: forall dom dataWidth . 
+  ( HiddenClockResetEnable dom, KnownNat dataWidth, 4 <= dataWidth )
+  => Circuit (PacketStream dom dataWidth EBHeader) (PacketStream dom dataWidth (Bool, RecordHeader))
+recordDepacketizerC = recordSplitterC |> depacketizerC metaMap
+  where
+    metaMap
+      :: (RecordHeader, Vec (dataWidth - 4) (BitVector 8))
+      -> Bool
+      -> (Bool, RecordHeader)
+    metaMap hdr b = (b, fst hdr)
+
+    -- This circuit splits a Etherbone @PacketStream@ into separate Record
+    -- streams. The @_last@ field indicates a end-of-record.
+    -- The metadata is replaced by a Bool indicating end-of-packet.
+    recordSplitterC = Circuit $ mealyB recordSplitterT 0
+    recordSplitterT
+      :: Unsigned 10
+      -> (Maybe (PacketStreamM2S dataWidth meta), PacketStreamS2M)
+      -> (Unsigned 10, (PacketStreamS2M, Maybe (PacketStreamM2S dataWidth Bool)))
+    recordSplitterT 0 (Nothing, _) = (0, (PacketStreamS2M True, Nothing))
+    recordSplitterT 0 (Just ps, oBwd) = (st', (oBwd, Just ps {_meta = False}))
+      where
+        wCount = bitCoerce $ resize $ _data ps !! (2 :: Integer)
+        rCount = bitCoerce $ resize $ _data ps !! (3 :: Integer)
```

Comment:
There is no performance related advantage, but `v at i `ensures `i <= length v` at compile time rather than producing an error at runtime if `i > length v`

Reply:
But if you have to jump through hoops to apply it I'm fine with leaving it

Reply:
I tried to do the hoops in the right way, not ad-hoc. One problem is that the constraint is of the form `m <= k` but `at` wants it of the form `m + k`. That we solve with `leToPlus`. But there's more, `at` doesn't just want `m + k` but in that expression, `k ~ n + 1`. I can't get `leToPlus` to do this for me, but it might be my Haskell fu is insufficient.

Reply:
The "proper" solution is this:

```diff
--- a/src/Clash/Cores/Etherbone/Base.hs
+++ b/src/Clash/Cores/Etherbone/Base.hs
@@ -208,8 +208,8 @@ recordDepacketizerC = recordSplitterC |> depacketizerC metaMap
     recordSplitterT 0 (Nothing, _) = (0, (PacketStreamS2M True, Nothing))
     recordSplitterT 0 (Just ps, oBwd) = (st', (oBwd, Just ps {_meta = False}))
       where
-        wCount = bitCoerce $ resize $ _data ps !! (2 :: Integer)
-        rCount = bitCoerce $ resize $ _data ps !! (3 :: Integer)
+        wCount = bitCoerce $ resize $ atLe d2 $ _data ps
+        rCount = bitCoerce $ resize $ atLe d3 $ _data ps
         wCount'
           | wCount > 0 = wCount + 1
           | otherwise  = 0
@@ -233,6 +233,9 @@ recordDepacketizerC = recordSplitterC |> depacketizerC metaMap
         oFwd = Just $ ps { _last = lst, _meta = isJust $ _last ps }
         lst = if left' == 0 then Just maxBound else Nothing
 
+atLe :: forall a k n . (k + 1) <= n => SNat k -> Vec n a -> a
+atLe = leToPlus @(k + 1) @n at
+
 -- | Helper function to convert a dataWidth in bytes to the correct format for
 -- the @_addrSize@ and @_portSize@ fields in the 'EBHeader'.
 sizeMask :: Integer -> BitVector 4
```

Reply:
Okay thanks, I'll keep `(!!)` for the time being then